pkg/storage/format/formatprovider.go

```diff
@@ -0,0 +1,131 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package format
+
+import (
+	"fmt"
+
+	"github.com/hyperledger/aries-framework-go/pkg/storage"
+)
+
+const (
+	failOpenUnderlyingStore      = "failed to open underlying store: %w"
+	failCloseUnderlyingStore     = "failed to close underlying store: %w"
+	failCloseAllUnderlyingStores = "failed to close all underlying stores: %w"
+	failFormatValue              = "failed to format value: %w"
+	failPutInUnderlyingStore     = "failed to put encrypted document in underlying store: %w"
+	failParseFormattedValue      = "failed to parse formatted value: %w"
+	failDeleteInUnderlyingStore  = "failed to delete key-value pair in underlying store: %w"
+)
+
+// Formatter represents a type that can encrypt and decrypt between
+// Structured Documents and Encrypted Documents.
+type Formatter interface {
+	FormatPair(k string, v []byte) ([]byte, error)
+	ParseValue([]byte) ([]byte, error)
+}
+
+// Provider is an encrypted storage provider that uses EDV document models
+// as defined in https://identity.foundation/secure-data-store/#data-model.
+// TODO (#2273): Generalize this to allow for other encrypted formats.
+type Provider struct {
```

Comment:
I forgot to include this in the original [comment](https://github.com/hyperledger/aries-framework-go/issues/2199#issuecomment-716617512), now updated:

> `storage.FormattedStore`: implements `storage.Store` and is composed of an underlying store and a `storage.Formatter`. It formats all data before saving it to the underlying store.

So this provider needs to be generalized and moved up a level into the `storage` package.

Reply:
Done. Note that I had to move the FormattedStore tests into their own package (formattedstore_test.go, still the same folder) in order to prevent a circular dependency between the memstore package and the store.storage package.